### PR TITLE
[FS] Add multisig withdrawal endpoint

### PR DESCRIPTION
**Pull request by rowandh**
* Add endpoint to build and sign a multisig tx when provided with the correct mnemonic/passphrases. Uses federation wallet for coin selection.
* Add the multisig address to the federation wallet info

### CODE REVIEW

File: src/Stratis.Features.FederatedPeg/Models/BuildMultisigTransactionRequest.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace Stratis.Features.FederatedPeg.Models
+{
+    /// <summary>
+    /// A class containing the necessary parameters for a build transaction request.
+    /// </summary>
+    public class BuildMultisigTransactionRequest
+    {
+        /// <summary>
+        /// A list of transaction recipients. For each recipient, specify the Pubkey script and the amount the
```
Comment (zeptin) [inline src/Stratis.Features.FederatedPeg/Models/BuildMultisigTransactionRequest.cs]: The recipient isn't really a pubkey, it's just any address considered valid by the current network. Minor nitpick

Comment (rowandh) [reply src/Stratis.Features.FederatedPeg/Models/BuildMultisigTransactionRequest.cs]: I copy-pasted this from the main wallet 😅

File: src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        /// Signs the transaction using the given private keys.
+        /// </summary>
+        /// <param name="recipients">The recipients of the transaction.</param>
+        /// <param name="privateKeys">The private keys of the multisig.</param>
+        /// <returns></returns>
+        public Transaction BuildTransaction(List<Recipient> recipients, Key[] privateKeys)
+        {
+            // FederationWalletTransactionHandler only supports signing with a single key - the fed wallet key - so we don't use it to build the transaction.
+            // However we still want to use it to determine what coins we need, so hack this together here to pass in to FederationWalletTransactionHandler.DetermineCoins.
+            var multiSigContext = new Wallet.TransactionBuildContext(recipients);
```
Comment (rowandh) [inline src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs]: Not a fan of doing this - but didn't want to modify the federation transaction builder to support multiple keys just yet.

File: src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+            (List<Coin> coins, List<Wallet.UnspentOutputReference> _) = FederationWalletTransactionHandler.DetermineCoins(this.federationWalletManager, this.network, multiSigContext, this.federatedPegSettings);
```
Comment (codingupastorm) [inline src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs]: You can remove this and the `AddCoins` - it only adds time to the process when there's a silly amount of UTXOs, and the `TransactionHandler` will get them in the same order anyway

File: src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+            // MultiSigAddress from the wallet is not safe. It's only pulled from multisig-wallet.json and can
+            // be different from the *actual* multisig address for the current redeem script.
+            // Instead, use the address from settings - it's derived from the redeem script provided at startup.
+            var multiSigAddress = this.federatedPegSettings.MultiSigAddress.ScriptPubKey;
```
Comment (codingupastorm) [inline src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs]: What if they put the wallet in the data folder and start it up without changing their given script. Then they will have the reverse effect of yours?

Should we put in a check that the wallet matches the given redeemscript? Seems best, could be a different PR.

File: src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs (hunk 3 above)
Comment (codingupastorm) [reply src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs]: Oh see what you've done now. You shouldn't have to do all those calculations for fees, just set the fee to be medium and put the recipient in without specifying the coins. Should be okay?

File: src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs (hunk 4 above)
Comment (rowandh) [reply src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs]: That's a risk too. I suggested exactly what you did on the DevOps issue, think it's a separate PR.

File: src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    {
+        private readonly IFederationWalletManager federationWalletManager;
+
+        private readonly IFederatedPegSettings federatedPegSettings;
+        
+        private readonly Network network;
+
+        private readonly IWalletFeePolicy walletFeePolicy;
+
+        public FedMultiSigWithdrawalTransactionBuilder(
```
Comment (codingupastorm) [inline src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs]: We already have a `WithdrawalTransactionBuilder`, maybe put something in the name to suggest this is for manual building and signing?

File: src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs (hunk 3 above)
Comment (codingupastorm) [reply src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs]: Nevermind, I see you're not using the TxHandler at all. Ignore these.

File: src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs (hunk 5 above)
Comment (zeptin) [reply src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs]: I agree, the naming should distinguish between the two more.

File: src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs (hunk 4 above)
Comment (zeptin) [reply src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs]: Since you have all the keys available you can make probably derive the redeem script within this code, although detecting if there's a difference between the wallet and settings is desirable as well I think.

Comment (zeptin) [reply src/Stratis.Features.FederatedPeg/FedMultiSigWithdrawalTransactionBuilder.cs]: Agreed that it can be a separate PR though.

File: src/Stratis.Features.FederatedPeg.Tests/Wallet/ManualMultiSigTransactionBuilderTests.cs [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        public ManualMultiSigTransactionBuilderTests()
+        {
+            this.network = new StratisTest();
+            this.federatedPegSettings = Substitute.For<IFederatedPegSettings>();
+            this.walletFeePolicy = Substitute.For<IWalletFeePolicy>();
+            this.multisigCoinSelector = Substitute.For<IMultisigCoinSelector>();
+        }
+
+        [Fact]
+        public void Multisig_Withdrawal_Transaction_Builds_Correctly()
```
Comment (codingupastorm) [inline src/Stratis.Features.FederatedPeg.Tests/Wallet/ManualMultiSigTransactionBuilderTests.cs]: 💯 

Could check tx is signed otherwise ace

**Comment by zeptin** [comment]
The tests look good to me, should be ready to merge
